src/jdk.httpserver/share/classes/sun/net/httpserver/SSLStreams.java

```diff
@@ -430,11 +430,16 @@ void doClosure () throws IOException {
             handshaking.lock();
             ByteBuffer tmp = allocate(BufType.APPLICATION);
             WrapperResult r;
+            Status st;
+            HandshakeStatus hs;
             do {
                 tmp.clear();
                 tmp.flip ();
                 r = wrapper.wrapAndSendX (tmp, true);
-            } while (r.result.getStatus() != Status.CLOSED);
+                hs = r.result.getHandshakeStatus();
+                st = r.result.getStatus();
+            } while (st != Status.CLOSED &&
+                        !(st == Status.OK && hs == HandshakeStatus.NOT_HANDSHAKING));
         } finally {
             handshaking.unlock();
         }
```

Comment:
Ok. The expectation is that doClosure() only involves sending a close acknowedged and not receiving anything. This is probably correct if doClosure() is called on reception of close_notify. In which case the change looks reasonable. I also see that a further expectation is that cycles of NEED_WRAP/NEED_UNWRAP only happen during the handshake, which is hopefully true.